File: day3/3.py
from timeit import default_timer as timer

def get_input():
    with open("./day3/3.in","r") as f:
        return [line.strip() for line in f.readlines()]


def p1():
    res, pos = 0, 0
    data = get_input()
    for line in data:
        if line[pos] == '#':
            res += 1
        pos = (pos + 3) % len(line)
    return res

def p2():
    slopes = [  (1, 1),
                (3, 1),
                (5, 1),
                (7, 1),
                (1, 2) ]
    data = get_input()
    res = 1
    for slope in slopes:
        posX, posY = 0, 0
        temp = 0 
        while posY < len(data):
            if data[posY][posX] == '#':
                temp += 1 
            posY = posY + slope[1]
            posX = (posX+slope[0]) % len(data[0])
        res *= temp
    return res

p1start = timer()
p1 = p1()
p1end = timer()
print(f"Part 1: {p1} in {p1end-p1start}s.")

p2start = timer()
p2 = p2()
p2end = timer()
print(f"Part 2: {p2} in {p2end-p2start}s.")

File: day3/3.go
package main

import (
	"fmt"
	"time"

	"../helpers"
)

func p1(data []string) {
	defer helpers.TimeTrack(time.Now())
	res, pos := 0, 0
	for _, line := range data {
		if line[pos] == '#' {
			res++
		}
		pos = (pos + 3) % len(line)
	}
	fmt.Printf("Part 1: %d ", res)
}

func p2(data []string) {
	defer helpers.TimeTrack(time.Now())
	res := 1
	slopes := [][]int{
		{1, 1},
		{3, 1},
		{5, 1},
		{7, 1},
		{1, 2},
	}
	for _, slope := range slopes {
		posX, posY := 0, 0
		temp := 0
		for posY < len(data) {
			if data[posY][posX] == '#' {
				temp++
			}
			posY = posY + slope[1]
			posX = (posX + slope[0]) % len(data[0])
		}
		res = res * temp
	}
	fmt.Printf("Part 2: %d ", res)
}

func main() {
	fmt.Println("Day 3: Toboggan Trajectory")
	fileName := "/Users/jack/Documents/Github/aoc-2020/day3/3.in"
	data, err := helpers.GetInputStringByLine(fileName)
	if err != nil {
		panic(err)
	}
	p1(data)
	p2(data)
}
